### ISSUE TITLE
Possible enhancement

### ISSUE THREAD

**Issue by leotiger** [getgrav/grav-plugin-form#271]
@rhukster: I continue working on the shoppingcart personalizer and I had a hard time to implementing file upload for dynamic page routes. Problem here: the ajax file upload only runs in settings from a kind of static blueprint, a page with a form in the frontmatter, etc. Doing dynamic stuff you normally don't want to save configurations. A real nice feature: if settings could be changed via an event hook.

I eventually prepare a PR to illustrate this better.

**Comment by leotiger** [comment]
I've created the Pull request #272 which adds two hooks:

* a hook to adapt the settings for uploaded files
* a hook which allows to manipulate the destination directory the very moment a file will be stored there, this is especially useful if you have to guarantee consistency of linked files in a given context.

**Comment by mahagr** [comment]
Latest admin already has support for image fields, though it is only possible if your plugin has its own save logic (or custom path in admin) extending `AdminBaseController`.

**Comment by leotiger** [comment]
@mahagr: That's not admin related, that's in the context of frontend. I have an own logic for ajax uploads in my plugin now as the current form implementation was not versatile enough. As the ajax response ends the lifecycle that's ok but I have to blow up the code with more than 200 lines of code to handle this.

Now, I could also run in a complete save stack but that's an overkill as I would in fact reproduce the form plugin by nearly 80%. As I have the need to clear existing files in the destination to assure that only the current uploads are available and associated with an object, in this case personalizations for bought products. As file uploads are handled via ajax and dorming in the session object, the best moment to assure this is when the files are moved from their temporary location into the destination. Right now I'm doing this when the form is created on the fly but that's a kick in the ass because I don't know if the form will be submitted or not but previous uploads get lost this way. My workaround for this is to allow personalizations only once and a administrator reset is needed to make the form available once again.

The PR offers a lot more flexibility and the opportunity to effect manipulations in the destination in the precise moment and circumstance.

**Comment by leotiger** [comment]
@mahagr Alternatively the files object could be exposed via an event hook. That's maybe even better as it would allow even more complex logic to be implemented without affecting the process of the form plugin.

**Comment by mahagr** [comment]
Fair enough, I didn't know your use case. I'm likely going to need something like this in very near future, so this is really something we need a proper solution.

All of this said -- I don't really like how Forms class is implemented and it may be a good time to try to figure out a better way to handle the whole form.

**Comment by leotiger** [comment]
I've exposed the queue itself now instead of destinations, far better.

**Comment by mahagr** [comment]
Yes, I like this new approach much better.
